### PR TITLE
[jit] Reduce special casing around 'training'

### LINKED ISSUE #26884: [jit] training attribute broken in recursive script
## 🐛 Bug

the model was scripted in eval mode and then saved with torch.jit.save. 
torch.jit.load loads the model in training mode. 
this is a problem, because in training mode dropouts are enabled, and also batchnorm needs to be adjusted. we expected torch.jit.load to load the model in the mode it was saved in. please fix this. 

## To Reproduce

**Steps to reproduce the behavior:**
```
tamas@super-duper-compute-machine:~/JUPYTER_LAB$ ipython
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.6.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch                                                                                                                              

In [2]: torch.__version__                                                                                                                         
Out[2]: '1.2.0'

In [3]: class MyCell(torch.nn.Module): 
   ...:     def __init__(self): 
   ...:         super(MyCell, self).__init__() 
   ...:         self.linear = torch.nn.Linear(4, 4) 
   ...:  
   ...:     def forward(self, x, h): 
   ...:         new_h = torch.tanh(self.linear(x) + h) 
   ...:         return new_h, new_h 
   ...:  
   ...: model = MyCell() 
   ...: x, h = torch.rand(3, 4), torch.rand(3, 4) 
   ...: scripted_model = torch.jit.script(model, (x, h))                                                                                          
/home/tamas/anaconda3/envs/cuda/lib/python3.7/site-packages/torch/jit/__init__.py:1158: UserWarning: `optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead
  warnings.warn("`optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead")

In [4]: scripted_model.code                                                                                                                       
Out[4]: 'def forward(self,\n    x: Tensor,\n    h: Tensor) -> Tuple[Tensor, Tensor]:\n  _0 = self.linear\n  _1 = _0.weight\n  _2 = _0.bias\n  if torch.eq(torch.dim(x), 2):\n    _3 = torch.__isnot__(_2, None)\n  else:\n    _3 = False\n  if _3:\n    bias = ops.prim.unchecked_unwrap_optional(_2)\n    ret = torch.addmm(bias, x, torch.t(_1), beta=1, alpha=1)\n  else:\n    output = torch.matmul(x, torch.t(_1))\n    if torch.__isnot__(_2, None):\n      bias0 = ops.prim.unchecked_unwrap_optional(_2)\n      output0 = torch.add_(output, bias0, alpha=1)\n    else:\n      output0 = output\n    ret = output0\n  new_h = torch.tanh(torch.add(ret, h, alpha=1))\n  return (new_h, new_h)\n'

In [5]: scripted_model.training                                                                                                                   
Out[5]: True

In [6]: scripted_model.eval()                                                                                                                     
Out[6]: 
WeakScriptModuleProxy(
  (linear): WeakScriptModuleProxy()
)

In [7]: scripted_model.training                                                                                                                   
Out[7]: False

In [8]: torch.jit.save(scripted_model, 'temp.pt')                                                                                                 

In [9]: m = torch.jit.load('temp.pt')                                                                                                             

In [10]: m.training                                                                                                                               
Out[10]: True
```

cc @ezyang @gchanan @zou3519 @suo

### PR DESCRIPTION
**Pull request by driazati**
Most of this was old cruft left over from special handling of `training` before we had a `bool` type. This makes all modules have a `training` attribute that is true by default and removes all other special handling.

Fixes #26884

Differential Revision: [D17728129](https://our.internmc.facebook.com/intern/diff/17728129/)

### CODE REVIEW

**Comment by suo** [comment]
Can you add a test case simulating the problem in that issue?

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -1592,6 +1586,9 @@ def __setattr__(self, attr, value):
                         raise RuntimeError("Could not register attribute '{}' of type '{}' for a value of type '{}'"
                                            .format(attr, value.type, type(value.value)))
                     return
+                if self._c._has_attribute(attr):
+                    attr_type = self._c._get_attribute_type(attr)
```
Comment (suo) [inline torch/jit/__init__.py]: we should use set_attribute here, not register_attribute

File: torch/csrc/jit/script/init.cpp [hunk 2]
```diff
@@ -546,6 +546,11 @@ void initJitScriptBindings(PyObject* module) {
       .def("_get_parameter", &Module::get_parameter)
       .def("_get_buffer", &Module::get_buffer)
       .def("_get_attribute", &Module::get_attribute)
+      .def("_get_attribute_type", [](Module& self, const std::string& attr) {
```
Comment (suo) [inline torch/csrc/jit/script/init.cpp]: shouldn't need this

File: torch/csrc/jit/script/module.h [hunk 3]
```diff
@@ -114,8 +114,14 @@ struct TORCH_API Module {
       std::shared_ptr<CompilationUnit> cu,
       bool shouldMangle = false);
   // module_value_ null and will be lazily initialized if is needed
-  Module() {}
-  Module(ModulePtr module_value) : module_value_(std::move(module_value)) {}
+Module() : Module(nullptr) {};
```
Comment (suo) [inline torch/csrc/jit/script/module.h]: Why is this here?

File: torch/csrc/jit/script/module.cpp [hunk 4]
```diff
@@ -33,15 +33,15 @@ static ModulePtr create_module_object(
 }
 
 Module::Module(c10::QualifiedName class_name)
-    : module_value_(create_module_object(
+    : Module(create_module_object(
```
Comment (suo) [inline torch/csrc/jit/script/module.cpp]: This is a weird construction because it sort of looks like `Module` inherits from `Module`. Can we just keep it the way it was?

File: torch/csrc/jit/script/module.cpp [hunk 5]
```diff
@@ -358,7 +358,7 @@ void Module::train(bool on) {
   if (auto slot = find_attribute("training")) {
     slot->setValue(on);
   } else {
-    register_attribute("training", BoolType::get(), on);
+    AT_ERROR("'training' attribute not found");
```
Comment (suo) [inline torch/csrc/jit/script/module.cpp]: we shouldn't use `AT_ERROR` in new code, it should either be `TORCH_INTERNAL_ASSERT` or `TORCH_CHECK` (if user facing)

**Review by suo — CHANGES_REQUESTED** [review]
It doesn't seem like this removes special casing for training, it just moves it around. Can we do the following:
`ScriptModule.__init__` calls `super(ScriptModule, self).__init__` and we remove the `Module._Module__construct` function entirely. That means that every ScriptModule will be initialized with a training attribute, which we just script normally.

In this world, every ScriptModule will now have a `training` attribute. You may have to fix up `__getstate__` and `__setstate__` in a bunch of places to make it work (look at the changes in https://github.com/pytorch/pytorch/pull/26666 is you want to know exactly where) but it should be fine otherwise.

File: torch/csrc/jit/script/module.cpp (hunk 4 above)
Comment (driazati) [reply torch/csrc/jit/script/module.cpp]: These changes were so that all the modules constructors would delegate to a single one that could register training, but that's not necessary in the latest commit

File: torch/csrc/jit/import.cpp [hunk 6]
```diff
@@ -42,6 +42,12 @@ void postSetStateValidate(const IValue& v) {
     // Verify that all the non-optional attributes have been initialized
     // TODO: Issue #20497
     if (attrType->kind() != TypeKind::OptionalType) {
+      if (slot.isNone() && attrName == "training") {
```
Comment (suo) [inline torch/csrc/jit/import.cpp]: I think this counts as another special case. I don't think it's correct to assume that anyone who defines a `__setstate__` will want their module to default to `eval()` (although it's probably true in most circumstances). 

We should just require that they initialize `self.training` in `__setstate__`, like any other attribute of the module. It's a little annoying, but more consistent overall.

File: torch/csrc/jit/script/module.h [hunk 7]
```diff
@@ -257,11 +257,7 @@ struct TORCH_API Module {
   }
   /// True if the module is in training mode.
   bool is_training() {
-    if (auto p = find_attribute("training")) {
-      return p->value().toBool();
-    }
-    // We are in training mode by default
-    return true;
+    return get_attribute("training").toBool();
```
Comment (suo) [inline torch/csrc/jit/script/module.h]: We can't remove this for BC reasons. Existing ScriptModules don't have "training" defined necessarily.

File: test/test_jit.py [hunk 8]
```diff
@@ -3442,6 +3442,27 @@ def forward(self, x):
         mod = torch.jit.script(MyMod())
         FileCheck().check_dag("NamedTuple").check_dag("Exception").run(mod.forward.graph)
 
+    def test_eval_python(self):
```
Comment (suo) [inline test/test_jit.py]: Can you add a test with the old API as well, for completeness.

File: torch/jit/__init__.py [hunk 9]
```diff
@@ -1513,7 +1513,6 @@ def __init__(self, optimize=None, _qualified_name=None, _compilation_unit=None,
                 self.__dict__['_c'] = torch._C.ScriptModule(_qualified_name, _compilation_unit, True)
 
             Module._Module__construct(self)
```
Comment (suo) [inline torch/jit/__init__.py]: The only point of this stupid `_module__construct` thing was to avoid setting "training". Let's just merge `__construct` and `__init__` in `nn.Module`. and call super's `__init__` here.

**Review by suo — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/import.cpp (hunk 6 above)
Comment (suo) [reply torch/csrc/jit/import.cpp]: You'll have to fix up places in our standard lib which do this now. As above, https://github.com/pytorch/pytorch/pull/26666 will tell you where you need to fix up, because I made the same change in that PR.

File: torch/jit/__init__.py (hunk 9 above)
Comment (driazati) [reply torch/jit/__init__.py]: What if someone makes a `script::Module` in C++? Should that also have a `training` attribute by default?

**Comment by suo** [comment]
you will need to fix test_quantization and test_mkldnn

File: test/test_jit.py [hunk 10]
```diff
@@ -3442,6 +3442,36 @@ def forward(self, x):
...
+        def _test(m):
+            self.assertTrue(m(torch.ones(2, 2)))
+            self.assertTrue(m.training)
+            self.assertTrue(m._c._get_attribute('training'))
+
+            m.eval()
+
+            self.assertFalse(m.training)
+            self.assertFalse(m._c._get_attribute('training'))
+            self.assertFalse(m(torch.ones(2, 2)))
```
Comment (suo) [inline test/test_jit.py]: Can we test that settings persist after a save and load as well?

**Review by suo — APPROVED** [review]
lgtm, still a bit hacky but this code will be heavily refactored soon(tm) anyway
